File: source/web/static/js/vas-core.js/api/labtrain.js
/**
 * [core/api/labtrain] - LabTrain API
 */
define(["core/api/interface", "liveq/LiveQ", "liveq/LabProtocol", "liveq/BufferReader", "core/config"], 

	function(APIInterface, LiveQ, LabProtocol, BufferReader, Config) {

		/**
		 * APISocket LabTrain
		 *
		 * @see {@link module:core/api/interface~APIInterface|APIInterface} (Parent class)
		 * @exports core/api/labtrain
		 */
		var APILabTrain = function(apiSocket, sequenceID, observables) {

			// Initialize superclass
			APIInterface.call(this, apiSocket);

			// Create an instance to LabProtocol and fetch all events
			this.labProtocol = new LabProtocol();
			this.labProtocol.forwardAllEventsTo(this);

			// Setup properties
			this.running = false;

			// Keep the sequence ID
			this.sequenceID = sequenceID;

			// Prepare handshake frame
			var handshakeFrame = {
				"version": LiveQ.version,
				"sequence": sequenceID,
				"observables": observables
			};

			// Do handshake
			this.sendAction("open", handshakeFrame);

			// We are connected
			this.connected = true;

			// Fire callbacks
			this.trigger('connected', this);

		}

		// Subclass from APIInterface
		APILabTrain.prototype = Object.create( APIInterface.prototype );

		/**
		 * Handle labSocket event
		 *
		 * @param {string} action - The action name
		 * @param {object} data - The action parameters
		 */
		APILabTrain.prototype.handleAction = function(action, data) {

			if (action == "status") {  /* Status message */
				console.log(data['message']);

				// Fire callbacks
				this.trigger('log', data['message'], data['vars']);

			} else if (action == "error") { /* Error message */
				console.error("I/O Error:",data['message']);

				// Fire callbacks
				this.trigger('error', data['message'], false);

			} else if (action == "completed") { /* Job completed */
				console.log("Job completed");

				// Fire callbacks
				this.trigger('completed');

				// Simulation is completed
				this.running = false;

			}
		}

		/**
		 * Handle labSocket binary frame
		 *
		 * @param {int} action - The action frame ID (16-bit integer)
		 * @param {ArrayBuffer} data - The action payload as a javascript ArrayBuffer
		 */
		APILabTrain.prototype.handleData = function(action, data) {

			// Encapsuate ArrayBuffer in a BufferReader class
			var reader = new BufferReader(data);

			// Skip the first 64-bit (frame header)
			reader.getUint32(); reader.getUint32();

			// Handle frame action
			if (action == 0x01) { /* Configuration Frame */

				// Handle configuration frame
				this.labProtocol.handleConfigFrame( reader );

			} else if (action == 0x02) { /* Histogram Data Frame */

				// Handle histogram data frame
				this.labProtocol.handleFrame( reader );

			}

		}

		/**
		 * Request training sequence to begin
		 */
		APILabTrain.prototype.beginTrain = function() {
			// Begin simulation with the given parameters
			this.sendAction("start");
			this.running = true;
		}

		/**
		 * Abort training sequence
		 */
		APILabTrain.prototype.stopTrain = function() {
			// Abort simulation
			this.sendAction("stop");
			this.running = false;
		}

		/**
		 * This event is fired when the socket is connected.
		 *
		 * @event module:core/api/labsocket~APILabTrain#connected		
		 */

		/**
		 * This event is fired when the socket is disconnected.
		 *
		 * @event module:core/api/labsocket~APILabTrain#disconnected		
		 */

		/**
		 * This event is fired when the socket is connected.
		 *
		 * @param {string} errorMessage - The error message
		 * @param {boolean} recoverable - If true the error is recoverable
		 * @event module:core/api/labsocket~APILabTrain#error		
		 */

		/**
		 * This event is fired when there was a simulation error.
		 *
		 * @param {string} errorMessage - The error message
		 * @event module:core/api/labsocket~APILabTrain#runError		
		 */

		/**
		 * This event is fired when the simulation is completed.
		 *
		 * @event module:core/api/labsocket~APILabTrain#completed		
		 */

		/**
		 * A log message arrived from the server.
		 *
		 * @param {string} logMessage - The message to log
		 * @event module:core/api/labsocket~APILabTrain#log		
		 */

		// Return the APILabTrain class
		return APILabTrain;

	}

);

File: liveq-webserver/webserver/h/api/labtrain.py

################################################################
# LiveQ - An interactive volunteering computing batch system
# Copyright (C) 2013 Ioannis Charalampidis
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
################################################################

import struct
import uuid
import logging
import base64
import datetime

import liveq.data.js as js
import liveq.data.histo.io as io
import tornado.escape

from liveq.data.histo.intermediate import IntermediateHistogramCollection
from liveq.data.histo.interpolate import InterpolatableCollection

from webserver.config import Config
from webserver.h.api import APIInterface

from tornado.ioloop import IOLoop
from webserver.common.training import OfflineSequence
from webserver.common.api import compileObservableHistoBuffers

class LabTrainInterface(APIInterface):

	def __init__(self, socket):
		"""
		Initialize the TRAIN SOCKET API interface
		"""
		APIInterface.__init__(self, socket, "labtrain", 0x02)

		# Setup local variables
		self.sequence = None
		self.running = False
		self.trainTimer = None

		# Tunable/Observable Trim
		self.observables = []

	####################################################################################
	# --------------------------------------------------------------------------------
	#                             API INTERFACE IMPLEMENTATION
	# --------------------------------------------------------------------------------
	####################################################################################

	def close(self):
		"""
		Cleanup logic when the channel is closed
		"""
		self.logger.info("Socket closed")

		# If we have a running tune, cancel it
		self.trainAbort()

	def ready(self):
		"""
		When socket is ready, get the user reference
		"""

		# Keep a local reference of the user
		self.user = self.socket.user


	def handleAction(self, action, param):
		"""
		Handle trainsocket actions
		"""
		
		# Process actions
		if action == "open":

			# Get client API version
			self.cversion = "0.1a"
			if 'version' in param:
				self.cversion = param['version']

			# Validate request
			if not 'sequence' in param:
				self.sendError("Missing parameter 'sequence'")
				return
			if not 'observables' in param:
				self.sendError("Missing parameter 'sequence'")
				return

			# We have a handshake with the agent.
			# Open offline sequence
			self.logger.info("Opening offline sequence '%s' with client v%s" % (param['sequence'], self.cversion))

			# Store histogram names
			self.observables = param['observables']

			# Open training sequence
			self.sequence = OfflineSequence( 
				baseDir=Config.TRAINSEQ_PATH,
				prefix="%s-" % param['sequence'],
				histoTrim=self.observables
				)

			# Send configuration frame
			self.sendConfigurationFrame()

		# Process actions
		elif action == "start":

			# Start the train sequence
			self.trainSequence()

		# Process actions
		elif action == "stop":

			# Abort the train sequence
			self.trainAbort()

		else:
			# Unknown request
			self.sendError("Unknown action '%s' requested" % action )

	####################################################################################
	# --------------------------------------------------------------------------------
	#                             TRAINING PLAYBACK SEQUENCE
	# --------------------------------------------------------------------------------
	####################################################################################


	def sendConfigurationFrame(self):
		"""
		Send the first, configuration frame to the agent
		"""

		# ==========================
		#  Histograms (Observables)
		# ==========================

		# Fetch descriptions for the histograms
		histoBuffers = compileObservableHistoBuffers( self.observables )

		# Compile buffer and send
		self.sendBuffer( 0x01, 
				# Header must be 64-bit aligned
				struct.pack("<BBHI", 
					2, 						# [8-bit]  Protocol
					0, 						# [8-bit]  Flags (1=FromInterpolation)
					0, 						# [16-bit] Number of events
					len(histoBuffers)		# [32-bit] Number of histograms
				) + ''.join(histoBuffers)
			)

	def trainSequence(self):
		"""
		Start training sequence
		"""

		# If we don't have sequence, exit
		if not self.sequence:
			return

		# Get next training histograms
		histos = self.sequence.next()
		if histos == None:
			# If we don't have other data, we are completed
			# Forward event to the user socket
			self.sendAction( "completed", {} )
			return

		# Pack histograms into javascript buffers
		histoBuffers = []
		for h in histos:
			# Pack buffers
			histoBuffers.append( js.packHistogram( h ) )

		# Compile buffer and send
		self.sendBuffer( 0x02, 
				# Header must be 64-bit aligned
				struct.pack("<BBHI", 
					2, 						# [8-bit]  Protocol
					0, 						# [8-bit]  Flags (1=FromInterpolation)
					0, 						# [16-bit] Number of events
					len(histoBuffers)		# [32-bit] Number of histograms
				) + ''.join(histoBuffers)
			)

		# Schedule next request
		self.trainTimer = IOLoop.instance().add_timeout(datetime.timedelta(0,0,0,100), self.trainSequence)

	def trainAbort(self):
		"""
		Abort a training sequence
		"""

		# Dispose sequence
		if self.sequence:
			self.sequence = None

		# Abort timer
		if self.trainTimer:
			IOLoop.instance().remove_timeout(self.trainTimer)
			self.trainTimer = None

